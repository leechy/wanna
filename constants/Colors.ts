/**
 * Below are the colors that are used in the app. The colors are defined in the light and dark mode.
 * There are many other ways to style your app. For example, [Nativewind](https://www.nativewind.dev/), [Tamagui](https://tamagui.dev/), [unistyles](https://reactnativeunistyles.vercel.app), etc.
 */

const tintColorLight = '#0a7ea4';
const tintColorDark = '#fff';

export const Colors = {
  light: {
    text: '#333333',
    background: '#f2f2f2',
    tint: tintColorLight,
    icon: '#687076',
    tabIconDefault: '#687076',
    tabIconSelected: tintColorLight,

    tabBarBackground: '#ffffff',
    smallButtonBackground: '#ffffff',

    primary: '#3a1fc1',
    barelyVisible: '#cccccc',
    disabled: '#bbbbbb',
    inactive: '#999999',
    touchable: '#666666',
    danger: '#e73d01',

    listBackground: '#ffffff',
    listSeparator: '#d6d6d6',

    inputText: '#000000',
    inputPlaceholder: '#999999',
    inputBackground: '#e3e3e4',
    cursorColor: '#3a1fc1',
    fullButtonText: '#000000',
    fullButtonBackground: '#FFD700',
  },
  dark: {
    text: '#cccccc',
    background: '#000000',
    tint: tintColorDark,
    icon: '#9BA1A6',
    tabIconDefault: '#9BA1A6',
    tabIconSelected: tintColorDark,

    tabBarBackground: '#151718',
    smallButtonBackground: '#151718',

    primary: '#ffd700',
    barelyVisible: '#333333',
    disabled: '#454545',
    inactive: '#666666',
    touchable: '#999999',
    danger: '#e73d01',

    listBackground: '#151718',
    listSeparator: '#0a0a0a',

    inputText: '#ffffff',
    inputPlaceholder: '#666666',
    inputBackground: '#151718',
    cursorColor: '#ffffff',
    fullButtonText: '#efefef',
    fullButtonBackground: '#3A1FC1',
  },
};

export const fullButtonBackgrounds = {
  light: ['#ffd700', '#ffe800', '#ffd700'],
  dark: ['#3a1fc1', '#2b00ff', '#3a1fc1'],
};
