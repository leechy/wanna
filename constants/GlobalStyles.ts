import { StyleSheet } from 'react-native';

export const globalStyles = StyleSheet.create({
  listItem: {
    flexDirection: 'row',
    alignItems: 'center',
    minHeight: 48,
    borderBottomWidth: 1,
    overflow: 'hidden',
  },
  listItemGradient: {
    flexDirection: 'row',
    alignItems: 'center',
    minHeight: 48,
  },
  listItemAction: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  listItemLeadButton: {
    padding: 5,
  },
  itemListLabelContainer: {
    flex: 1,
    flexDirection: 'column',
    alignItems: 'flex-start',
    justifyContent: 'space-between',
    paddingVertical: 8,
  },
  itemListlabel: {
    fontFamily: 'Montserrat',
    fontSize: 15,
    lineHeight: 18,
  },
  itemListLabelMetaContainer: {
    paddingTop: 1,
    flexDirection: 'row',
    alignItems: 'center',
    gap: 8,
    marginStart: -2,
    opacity: 0.6,
  },
  itemListLabelMeta: {
    fontFamily: 'Montserrat',
    fontSize: 11,
  },
  itemListLabelMetaItem: {
    flexDirection: 'row',
    alignItems: 'center',
    gap: 2,
  },
  itemListProgress: {
    fontFamily: 'Montserrat',
    fontSize: 12,
    lineHeight: 14,
    textAlign: 'center',
    minWidth: 40,
  },
  titleContainer: {
    paddingHorizontal: 16,
  },
});
