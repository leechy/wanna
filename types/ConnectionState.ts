export type ConnectionState = {
  isConnected: boolean;
  lastConnected: string | null;
  errors: string[];
};
