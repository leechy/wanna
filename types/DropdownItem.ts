import { SvgProps } from 'react-native-svg';

export type DropdownItem = {
  label: string;
  onPress?: (item?: DropdownItem) => void;
  icon?: React.FC<SvgProps>;
  color?: string;
};
