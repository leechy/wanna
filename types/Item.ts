export type ItemUser = {
  uid: string;
  names: string;
};

export type Item = {
  itemId: string;
  listItemId: string;
  listId: string;

  name: string;
  type: 'item' | 'task';
  units?: string;
  quantity: string | number;
  deadline?: string;

  active?: boolean;
  public?: boolean;

  ongoing?: boolean;
  assignee?: string | null;
  assigneeId?: string | null;
  completed?: boolean;
  completedAt?: string | null;

  sortOrder?: number;

  deleted?: boolean;
  createdAt?: string;
  updatedAt?: string;
};
