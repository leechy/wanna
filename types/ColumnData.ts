import { SvgProps } from 'react-native-svg';
import { ListItem } from '@/types/listItem';

export type ColumnData = {
  title: string;
  color?: string;
  action?: React.ReactNode;
  newItemLabel?: string;
  newItemHandler?: () => void;
  newItemIcon?: React.FC<SvgProps>;
  actionHandler?: (item: ListItem) => void;
  checkboxHandler?: (item: ListItem) => void;
  longPressHandler?: (item: ListItem) => void;
  editHandler?: (item: ListItem) => void;
  deleteHandler?: (item: ListItem) => void;
  resetHandler?: (item: ListItem) => void;
  items: ListItem[];
  onToggle?: () => void;
  isOpen?: boolean;
  showEmpty?: boolean;
  emptyText?: string;
  onNew?: () => void;
};
