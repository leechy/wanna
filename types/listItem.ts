import { Item } from '@/types/Item';

export type ListItem = {
  type: // item types
  | 'new'
    | 'item'
    | 'task'
    // list types
    | 'shopping-list'
    | 'project'
    | 'recipe'
    // other types
    | 'contact'
    | 'group';
  id: string;
  label: string;
  list?: string;
  shared?: string[];
  deadline?: string | number | null;
  quantity?: number;
  ongoing?: boolean | number; // replacement for the progress property
  completed?: boolean | number;
  updatedAt?: string;
  sortOrder?: number;
  completedAt?: string | null;
  assignee?: string | null;
  assigneeId?: string | null;
};

export function convertItemsToListItems(items?: Item[] | null): ListItem[] {
  if (!items) {
    return [];
  }

  return items.map((item) => ({
    type: 'item',
    id: item.listItemId,
    label: item.name,
    listId: item.listId,
    deadline: item.deadline,
    quantity: item.quantity,
    ongoing: item.ongoing,
    assignee: item.assignee,
    assigneeId: item.assigneeId,
    completed: item.completed,
    completedAt: item.completedAt,
    updatedAt: item.updatedAt,
    sortOrder: item.sortOrder,
  }));
}
