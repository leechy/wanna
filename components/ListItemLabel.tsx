// hooks
import { useThemeColor } from '@/hooks/useThemeColor';

// components
import { Text, View } from 'react-native';

import CalendarIcon from '@/assets/symbols/square-calendar.svg';
import ListIcon from '@/assets/symbols/square-list.svg';
import PersonaIcon from '@/assets/symbols/persona.svg';
import TwoPersonasIcon from '@/assets/symbols/two-personas.svg';
import ThreePersonasIcon from '@/assets/symbols/three-personas.svg';

// utils
import { humanDate } from '@/utils/dates';
import { globalStyles } from '@/constants/GlobalStyles';

// types
import { ListItem } from '@/types/listItem';

interface ListItemLabelProps {
  item: ListItem;
  showQuantity?: boolean;
}

export default function ListItemLabel({ item, showQuantity }: ListItemLabelProps) {
  const textColor = useThemeColor({}, 'text');
  const touchableColor = useThemeColor({}, 'touchable');
  const dangerColor = useThemeColor({}, 'danger');

  const overdue = (item.deadline || 0) < Date.now();

  return (
    <View style={globalStyles.itemListLabelContainer}>
      <Text style={[globalStyles.itemListlabel, { color: textColor }]} adjustsFontSizeToFit={true} numberOfLines={2}>
        {showQuantity && (item.quantity || 0) > 1 && <>{item.quantity} &times; </>}
        {item.label}
      </Text>
      {(item.deadline || item.list || (item.shared?.length || 0) > 0) && (
        <View style={globalStyles.itemListLabelMetaContainer}>
          {item.deadline && (
            <View style={globalStyles.itemListLabelMetaItem}>
              <CalendarIcon width={14} height={14} color={overdue ? dangerColor : touchableColor} />
              <Text style={[globalStyles.itemListLabelMeta, { color: overdue ? dangerColor : touchableColor }]}>
                {humanDate(item.deadline)}
              </Text>
            </View>
          )}

          {item.list && (
            <View style={globalStyles.itemListLabelMetaItem}>
              <ListIcon width={14} height={14} color={touchableColor} />
              <Text style={[globalStyles.itemListLabelMeta, { color: touchableColor }]}>{item.list}</Text>
            </View>
          )}

          {(item.shared?.length || 0) > 0 && (
            <View style={globalStyles.itemListLabelMetaItem}>
              {item.shared!.length === 1 ? (
                <>
                  <PersonaIcon width={14} height={14} color={touchableColor} />
                  <Text style={[globalStyles.itemListLabelMeta, { color: touchableColor }]}>{item.shared![0]}</Text>
                </>
              ) : item.shared!.length === 2 ? (
                <>
                  <TwoPersonasIcon width={14} height={14} color={touchableColor} />
                  <Text style={[globalStyles.itemListLabelMeta, { color: touchableColor }]}>
                    {item.shared!.join(' and ')}
                  </Text>
                </>
              ) : (
                <>
                  <ThreePersonasIcon width={14} height={14} color={touchableColor} />
                  <Text style={[globalStyles.itemListLabelMeta, { color: touchableColor }]}>
                    {item.shared!.length} people
                  </Text>
                </>
              )}
            </View>
          )}
        </View>
      )}
    </View>
  );
}
