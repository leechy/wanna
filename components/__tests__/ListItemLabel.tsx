import React from 'react';
import { render } from '@testing-library/react-native';
import ListItemLabel from '../ListItemLabel';
import { ListItem } from '@/types/listItem';
import { Colors } from '@/constants/Colors';

jest.mock('../../assets/symbols/square-calendar.svg', () => 'CalendarIcon');
jest.mock('../../assets/symbols/square-list.svg', () => 'ListIcon');
jest.mock('../../assets/symbols/persona.svg', () => 'PersonaIcon');
jest.mock('../../assets/symbols/two-personas.svg', () => 'TwoPersonasIcon');
jest.mock('../../assets/symbols/three-personas.svg', () => 'ThreePersonasIcon');

// Mock the humanDate function
jest.mock('../../utils/dates', () => ({
  humanDate: jest.fn(),
}));

describe('ListItemLabel', () => {
  it('renders the label correctly', () => {
    const item: ListItem = { id: '0', type: 'item', label: 'Test Item', quantity: 3 };
    const { getByText } = render(<ListItemLabel item={item} showQuantity={false} />);
    expect(getByText('Test Item')).toBeTruthy();
  });

  it('shows the quantity when showQuantity is true', () => {
    const item: ListItem = { id: '1', type: 'item', label: 'Test Item', quantity: 3 };
    const { getByText } = render(<ListItemLabel item={item} showQuantity={true} />);
    expect(getByText('3 Ã— Test Item')).toBeTruthy();
  });

  it('applies the correct styles', () => {
    const item: ListItem = { id: '3', type: 'item', label: 'Test Item' };
    const { getByText } = render(<ListItemLabel item={item} showQuantity={false} />);
    const label = getByText('Test Item');
    expect(label.props.style).toHaveProperty('[1].color', '#333333');
  });

  it('shows the deadline if set', () => {
    const item: ListItem = { id: '2', type: 'item', label: 'Test Item', deadline: Date.now() + 10000 };
    const { getByTestId } = render(
      <ListItemLabel item={item} showQuantity={false} testCalendarID="item-label-calendar" />
    );
    expect(getByTestId('item-label-calendar')).toBeTruthy();
  });

  it('applies the danger color when overdue', () => {
    const item: ListItem = { id: '4', type: 'item', label: 'Test Item', deadline: Date.now() - 10000 };
    const { getByTestId } = render(
      <ListItemLabel item={item} showQuantity={false} testCalendarID="item-label-calendar" />
    );
    const date = getByTestId('item-label-calendar');
    expect(date.children[0]).toHaveProperty('type', 'CalendarIcon');
    expect(date.children[1].props.style).toHaveProperty('[1].color', Colors['light'].danger);
  });

  it('shows the list if set', () => {
    const item: ListItem = { id: '2', type: 'item', label: 'Test Item', list: 'test list' };
    const { getByText, getByTestId } = render(
      <ListItemLabel item={item} showQuantity={false} testListID="item-label-list" />
    );
    const list = getByTestId('item-label-list');
    expect(list).toBeTruthy();
    expect(list.children[0]).toHaveProperty('type', 'ListIcon');
    expect(getByText('test list')).toBeTruthy();
  });

  it('shows the one person correctly', () => {
    const item: ListItem = { id: '2', type: 'item', label: 'Test Item', shared: ['user1'] };
    const { getByText, getByTestId } = render(
      <ListItemLabel item={item} showQuantity={false} testPersonasID="item-label-personas" />
    );
    const shared = getByTestId('item-label-personas');
    expect(shared).toBeTruthy();
    expect(getByText('user1')).toBeTruthy();
    expect(shared.children[0]).toHaveProperty('type', 'PersonaIcon');
  });

  it('shows the two persons correctly', () => {
    const item: ListItem = { id: '2', type: 'item', label: 'Test Item', shared: ['user1', 'user2'] };
    const { getByText, getByTestId } = render(
      <ListItemLabel item={item} showQuantity={false} testPersonasID="item-label-personas" />
    );
    expect(getByText('user1 and user2')).toBeTruthy();
    expect(getByTestId('item-label-personas').children[0]).toHaveProperty('type', 'TwoPersonasIcon');
  });

  it('shows the more than 2 persons correctly', () => {
    const item: ListItem = { id: '2', type: 'item', label: 'Test Item', shared: ['user1', 'user2', 'user3', 'user4'] };
    const { getByText, getByTestId } = render(
      <ListItemLabel item={item} showQuantity={false} testPersonasID="item-label-personas" />
    );
    expect(getByText('4 people')).toBeTruthy();
    expect(getByTestId('item-label-personas').children[0]).toHaveProperty('type', 'ThreePersonasIcon');
  });

  it('matches snapshot', () => {
    const item: ListItem = { id: '5', type: 'item', label: 'Test Item', quantity: 3, deadline: Date.now() + 10000 };
    const { toJSON } = render(<ListItemLabel item={item} showQuantity={true} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
