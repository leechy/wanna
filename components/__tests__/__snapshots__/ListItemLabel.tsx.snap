// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItemLabel matches snapshot 1`] = `
<View
  style={
    {
      "alignItems": "flex-start",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "space-between",
      "paddingVertical": 8,
    }
  }
>
  <Text
    adjustsFontSizeToFit={false}
    numberOfLines={2}
    style={
      [
        {
          "fontFamily": "Montserrat",
          "fontSize": 15,
          "lineHeight": 18,
        },
        {
          "color": "#333333",
        },
      ]
    }
  >
    3
     Ã— 
    Test Item
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "gap": 8,
        "marginStart": -2,
        "opacity": 0.6,
        "paddingTop": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 2,
        }
      }
    >
      <CalendarIcon
        color="#666666"
        height={14}
        width={14}
      />
      <Text
        style={
          [
            {
              "fontFamily": "Montserrat",
              "fontSize": 11,
            },
            {
              "color": "#666666",
            },
          ]
        }
      />
    </View>
  </View>
</View>
`;
